spring-cloud-dataflow-container-registry/src/main/java/org/springframework/cloud/dataflow/container/registry/ContainerImageRestTemplateFactory.java

```diff
@@ -86,6 +77,12 @@
  */
 public class ContainerImageRestTemplateFactory {
 
+	private static final String CUSTOM_REGISTRY = "custom-registry";
+	private static final String AMZ_CREDENTIAL = "X-Amz-Credential";
+	private static final String AUTHORIZATION_HEADER = "Authorization";
+	private static final String AZURECR_URI_SUFFIX = "azurecr.io";
+	private static final String BASIC_AUTH = "Basic";
+
 	private final RestTemplateBuilder restTemplateBuilder;
 
 	private final ContainerRegistryProperties properties;
```

Comment:
**[UNRELATED]** This was already in this class, but we can 
- use Java 17 records to reduce this CacheKey bolierplate
- use ConcurrentHashMap properly w/ `computeIfAbsent`
to shrink this down into a few lines.

The code in question is here: https://github.com/spring-cloud/spring-cloud-dataflow/pull/5996/files#diff-d659fdfefb699e093be40defb784f08e85441f3a595328ee597aa899184ef6b4R97-R151

The code would look like this once refactored:
```java
	/**
	 * Depends on the disablesSslVerification and useHttpProxy a 4 different RestTemplate configurations might be
	 * used at the same time for interacting with different container registries.
	 * The cache map allows reusing the RestTemplates for given useHttpProxy and disablesSslVerification combination.
	 */
	private final ConcurrentHashMap<CacheKey, RestTemplate> restTemplateCache;

	/**
	 * Unique key to cache rest templates by for any skipSslVerification / withHttpProxy combination.
	 */
	private record CacheKey(boolean skipSslVerification, boolean withHttpProxy) {
	}

	public ContainerImageRestTemplateFactory(RestTemplateBuilder restTemplateBuilder, ContainerRegistryProperties properties) {
		this.restTemplateBuilder = restTemplateBuilder;
		this.properties = properties;
		this.restTemplateCache = new ConcurrentHashMap<>();
	}

	public RestTemplate getContainerRestTemplate(boolean skipSslVerification, boolean withHttpProxy) {
		return this.getContainerRestTemplate(skipSslVerification, withHttpProxy, Collections.emptyMap());
	}

	public RestTemplate getContainerRestTemplate(boolean skipSslVerification, boolean withHttpProxy, Map<String, String> extra) {
		var cacheKey = new CacheKey(skipSslVerification, withHttpProxy);
		try {
			return this.restTemplateCache.computeIfAbsent(cacheKey,
					(k) -> createContainerRestTemplate(k.skipSslVerification(), k.withHttpProxy(), extra));
		}
		catch (Exception ex) {
			throw new ContainerRegistryException("Failed to create REST template for (%s)".formatted(cacheKey), ex);
		}
	}
```

If you add this just do it in a separate commit. I am also glad to do it in a polish commit during merge.

Reply:
done

Reply:
I also decided to add `extra` to the cache key. If different values are used for extra the behaviour of the RestTemplate may not match expectation if the cached entry wasn't created with same values in `extra`